src/debugging/DockerServerReadyAction.ts

```diff
@@ -213,54 +190,187 @@ export class ServerReadyDetector extends vscode.Disposable {
+            && configuration.dockerOptions.dockerServerReadyAction) {
+            let tracker = DockerDebugAdapterTrackerFactory.trackers.get(session);
+            if (!tracker) {
+                tracker = new MultiOutputDockerServerReadyManager(session);
+                DockerDebugAdapterTrackerFactory.trackers.set(session, tracker);
             }
-            return debugConfiguration;
+
+            return tracker.tracker;
         }
-    }));
+
+        return undefined;
+    }
+
+    public static stop(session: vscode.DebugSession): void {
+        const tracker = DockerDebugAdapterTrackerFactory.trackers.get(session);
+
+        if (tracker) {
+            DockerDebugAdapterTrackerFactory.trackers.delete(session);
+            tracker.dispose();
+        }
+    }
+
+    public createDebugAdapterTracker(session: vscode.DebugSession): vscode.ProviderResult<vscode.DebugAdapterTracker> {
+        return DockerDebugAdapterTrackerFactory.start(session);
+    }
 }
 
-function startTrackerForType(context: vscode.ExtensionContext, type: string): void {
-
-    // scan debug console output for a PORT message
-    context.subscriptions.push(vscode.debug.registerDebugAdapterTrackerFactory(type, {
-        createDebugAdapterTracker(session: vscode.DebugSession): vscode.DebugAdapterTracker {
-            const detector = ServerReadyDetector.start(session);
-            if (detector) {
-                return {
-                    // tslint:disable: no-unsafe-any
-                    onDidSendMessage: m => {
-                        if (m.type === 'event' && m.event === 'output' && m.body) {
-                            switch (m.body.category) {
-                                case 'console':
-                                case 'stderr':
-                                case 'stdout':
-                                    if (m.body.output) {
-                                        detector.detectPattern(m.body.output);
-                                    }
-                                    break;
-                                default:
-                            }
-                        }
-                    }
-                    // tslint:enable: no-unsafe-any
-                };
+class DockerServerReadyDebugConfigurationProvider implements vscode.DebugConfigurationProvider {
+    private readonly trackers: Set<string> = new Set<string>();
+    private readonly trackerFactory: DockerDebugAdapterTrackerFactory = new DockerDebugAdapterTrackerFactory();
+
+    public constructor(private readonly context: vscode.ExtensionContext) {
+    }
+
+    public resolveDebugConfiguration?(folder: vscode.WorkspaceFolder | undefined, debugConfiguration: ResolvedDebugConfiguration, token?: vscode.CancellationToken): vscode.ProviderResult<vscode.DebugConfiguration> {
+        if (debugConfiguration && debugConfiguration.type && debugConfiguration.dockerOptions && debugConfiguration.dockerOptions.dockerServerReadyAction) {
```

Comment:
We should be able to now consolidate these long expressions (here and elsewhere), e.g.:

`if (debugConfiguration?.type && debugConfiguration?.dockerOptions?.dockerServerReadyAction) {`